go.mod

```diff
@@ -0,0 +1,143 @@
+	cloud.google.com/go/storage v1.16.1 // indirect
+	code.gitea.io/sdk/gitea v0.15.1 // indirect
+	github.com/AlekSi/pointer v1.2.0 // indirect
+	github.com/Azure/azure-pipeline-go v0.2.3 // indirect
+	github.com/Azure/azure-sdk-for-go v57.0.0+incompatible // indirect
+	github.com/Azure/azure-storage-blob-go v0.14.0 // indirect
+	github.com/Azure/go-autorest v14.2.0+incompatible // indirect
+	github.com/Azure/go-autorest/autorest v0.11.20 // indirect
+	github.com/Azure/go-autorest/autorest/adal v0.9.15 // indirect
+	github.com/Azure/go-autorest/autorest/azure/auth v0.5.8 // indirect
+	github.com/Azure/go-autorest/autorest/azure/cli v0.4.3 // indirect
+	github.com/Azure/go-autorest/autorest/date v0.3.0 // indirect
+	github.com/Azure/go-autorest/autorest/to v0.4.0 // indirect
+	github.com/Azure/go-autorest/autorest/validation v0.3.1 // indirect
+	github.com/Azure/go-autorest/logger v0.2.1 // indirect
+	github.com/Azure/go-autorest/tracing v0.6.0 // indirect
+	github.com/DisgoOrg/disgohook v1.4.4 // indirect
+	github.com/DisgoOrg/log v1.1.0 // indirect
+	github.com/DisgoOrg/restclient v1.2.8 // indirect
+	github.com/Masterminds/goutils v1.1.1 // indirect
+	github.com/Masterminds/semver v1.5.0 // indirect
+	github.com/Masterminds/semver/v3 v3.1.1 // indirect
+	github.com/Masterminds/sprig v2.22.0+incompatible // indirect
+	github.com/Microsoft/go-winio v0.5.0 // indirect
+	github.com/ProtonMail/go-crypto v0.0.0-20210512092938-c05353c2d58c // indirect
+	github.com/acomagu/bufpipe v1.0.3 // indirect
+	github.com/atc0005/go-teams-notify/v2 v2.6.1 // indirect
+	github.com/aws/aws-sdk-go v1.40.34 // indirect
+	github.com/aws/aws-sdk-go-v2 v1.9.0 // indirect
+	github.com/aws/aws-sdk-go-v2/config v1.7.0 // indirect
+	github.com/aws/aws-sdk-go-v2/credentials v1.4.0 // indirect
+	github.com/aws/aws-sdk-go-v2/feature/ec2/imds v1.5.0 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/ini v1.2.2 // indirect
+	github.com/aws/aws-sdk-go-v2/service/internal/presigned-url v1.3.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/kms v1.5.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/sso v1.4.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/sts v1.7.0 // indirect
+	github.com/aws/smithy-go v1.8.0 // indirect
+	github.com/aymanbagabas/go-osc52 v1.0.3 // indirect
+	github.com/blakesmith/ar v0.0.0-20190502131153-809d4375e1fb // indirect
+	github.com/caarlos0/ctrlc v1.1.0 // indirect
+	github.com/caarlos0/env/v6 v6.9.3 // indirect
+	github.com/caarlos0/go-reddit/v3 v3.0.1 // indirect
+	github.com/caarlos0/go-shellwords v1.0.12 // indirect
+	github.com/caarlos0/log v0.1.1 // indirect
+	github.com/cavaliergopher/cpio v1.0.1 // indirect
+	github.com/cenkalti/backoff/v4 v4.1.2 // indirect
+	github.com/charmbracelet/lipgloss v0.5.1-0.20220615005615-2e17a8a06096 // indirect
+	github.com/cpuguy83/go-md2man/v2 v2.0.2 // indirect
+	github.com/dghubble/go-twitter v0.0.0-20211115160449-93a8679adecb // indirect
+	github.com/dghubble/oauth1 v0.7.1 // indirect
```

Comment:
May be a lot of stuff from other test you did?
I don't know well the go.mod behavior.
May it exists a clean stuff or rebuild dependencies?

Reply:
ah, was awful (I'm investigating) I think all is coming from goreleaser as a go dependency?

```
sylvain@lap43: ~/code/go/src/github.com/docopt/docopts$ make install_builddep
go mod tidy
go: downloading github.com/goreleaser/goreleaser v1.10.3
go: downloading github.com/caarlos0/log v0.1.1
go: downloading github.com/caarlos0/ctrlc v1.1.0
go: downloading github.com/charmbracelet/lipgloss v0.5.1-0.20220615005615-2e17a8a06096
go: downloading github.com/invopop/jsonschema v0.5.0
go: downloading github.com/muesli/mango-cobra v1.2.0
go: downloading github.com/muesli/roff v0.1.0
go: downloading github.com/spf13/cobra v1.5.0
go: downloading github.com/muesli/termenv v0.12.1-0.20220615005108-4e9068de9898
go: downloading golang.org/x/sync v0.0.0-20220513210516-0976fa681c29
go: downloading github.com/stretchr/testify v1.8.0
go: downloading github.com/matryer/is v1.4.0
go: downloading github.com/goreleaser/nfpm/v2 v2.16.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/muesli/reflow v0.2.1-0.20210115123740-9e1d0d53df68
go: downloading github.com/iancoleman/orderedmap v0.2.0
go: downloading github.com/muesli/mango v0.1.0
go: downloading github.com/muesli/mango-pflag v0.1.0
go: downloading github.com/cpuguy83/go-md2man/v2 v2.0.2
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/aymanbagabas/go-osc52 v1.0.3
go: downloading golang.org/x/sys v0.0.0-20220209214540-3681064d5158
go: downloading code.gitea.io/sdk/gitea v0.15.1
go: downloading github.com/google/go-github/v45 v45.2.0
go: downloading github.com/xanzy/go-gitlab v0.68.2
go: downloading golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5
go: downloading github.com/jarcoal/httpmock v1.2.0
go: downloading github.com/Masterminds/semver/v3 v3.1.1
go: downloading golang.org/x/crypto v0.0.0-20220307211146-efcb8507fb70
go: downloading github.com/Masterminds/semver v1.5.0
go: downloading github.com/charmbracelet/keygen v0.3.0
go: downloading github.com/caarlos0/go-shellwords v1.0.12
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/imdario/mergo v0.3.13
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/russross/blackfriday/v2 v2.1.0
go: downloading gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c
go: downloading github.com/hashicorp/go-version v1.2.1
go: downloading github.com/google/go-querystring v1.1.0
go: downloading github.com/google/go-cmp v0.5.8
go: downloading github.com/hashicorp/go-cleanhttp v0.5.2
go: downloading github.com/hashicorp/go-retryablehttp v0.7.1
go: downloading golang.org/x/time v0.0.0-20220411224347-583f2d630306
go: downloading gocloud.dev v0.24.0
go: downloading github.com/DisgoOrg/disgohook v1.4.4
go: downloading github.com/caarlos0/env/v6 v6.9.3
go: downloading github.com/caarlos0/go-reddit/v3 v3.0.1
go: downloading github.com/slack-go/slack v0.11.0
go: downloading gopkg.in/mail.v2 v2.3.1
go: downloading github.com/atc0005/go-teams-notify/v2 v2.6.1
go: downloading github.com/go-telegram-bot-api/telegram-bot-api v4.6.4+incompatible
go: downloading github.com/dghubble/go-twitter v0.0.0-20211115160449-93a8679adecb
go: downloading github.com/dghubble/oauth1 v0.7.1
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/goreleaser/fileglob v1.3.0
go: downloading github.com/caarlos0/sshmarshal v0.0.0-20220308164159-9ddb9f83c6b3
go: downloading github.com/AlekSi/pointer v1.2.0
go: downloading github.com/goreleaser/chglog v0.1.2
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/blakesmith/ar v0.0.0-20190502131153-809d4375e1fb
go: downloading github.com/ulikunitz/xz v0.5.10
go: downloading github.com/xi2/xz v0.0.0-20171230120015-48954b6210f8
go: downloading github.com/google/rpmpack v0.0.0-20220314092521-38642b5e571e
go: downloading github.com/ProtonMail/go-crypto v0.0.0-20210512092938-c05353c2d58c
go: downloading github.com/caarlos0/go-rpmutils v0.2.1-0.20211112020245-2cd62ff89b11
go: downloading github.com/kr/pretty v0.2.1
go: downloading github.com/hashicorp/go-hclog v0.9.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.3
go: downloading github.com/Azure/azure-storage-blob-go v0.14.0
go: downloading github.com/Azure/go-autorest/autorest/adal v0.9.15
go: downloading github.com/Azure/go-autorest/autorest v0.11.20
go: downloading github.com/Azure/go-autorest v14.2.0+incompatible
go: downloading github.com/google/wire v0.5.0
go: downloading github.com/Azure/go-autorest/autorest/azure/auth v0.5.8
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/aws/aws-sdk-go v1.40.34
go: downloading cloud.google.com/go v0.94.0
go: downloading cloud.google.com/go/storage v1.16.1
go: downloading github.com/googleapis/gax-go/v2 v2.1.0
go: downloading google.golang.org/api v0.56.0
^Cmake: *** [Makefile:14: install_builddep] Interrupt
```

Rebuilding go.mod doesn't help:
https://stackoverflow.com/questions/65512353/how-to-undo-go-mod-init

```
rm go.mod
go mod init
go mod tidy
[...] plenty of stuff bloating my filesystem
```

So lets try it (173 dep downloaded) :thinking: 

nope a step is missing about installing goreleaser

```
sylvain@lap43: ~/code/go/src/github.com/docopt/docopts$ make clean
rm -f docopts-* docopts README.tmp dist/*
sylvain@lap43: ~/code/go/src/github.com/docopt/docopts$ make 
GOVERSION=$(go version) goreleaser build --rm-dist --snapshot --single-target -o docopts
/bin/sh: 1: goreleaser: not found
make: *** [Makefile:11: docopts] Error 127
```

```

Reply:
Oh sorry I thought you'd be familiar with Go modules so I didn't explain that.

Yea so these are all dependencies of `goreleaser` (or dependencies of its dependencies, and so on).
The `go.mod` file doesn't just contain direct dependencies, but also _indirect_ dependencies.

Note that in the `go.mod` file, there's actually a comment saying `// indirect` and the indirect deps are split into a separate `require` block as well.

That's all automatically generated too by `go mod tidy`. The `tidy` part will remove any dependencies that aren't necessary as well, so there's nothing extra.

`goreleaser` can do a ton of things, so unfortunately with that comes a bunch of deps. The functionality here is only a small part of what it can do, so a lot of those deps probably aren't used too.

Just to be clear -- these are all development dependencies just for using `goreleaser` (that's what `tools.go` is, a list of dev deps basically), so none of these make it into the actual `docopts` binary, as they're not used at compile-time.

Reply:
## none of the hundred of downloaded lib make it into the actual docopts binary

Yes, that's what I think too. But are they mandatory to be downloaded just for the purpose of running `goreleaser`?

> Oh sorry I thought you'd be familiar with Go modules so I didn't explain that.

In fact, I started coding Go as a casual coder transposing my experience from other language. But at the time I started Go module was not made that way or not at all. As you figure out probably, `docopts` is a small code with very low external lib.

So yes I'm not so familiar with module. I will learn.

But I suppose, that they're probably not needed to be listed as part of the project?
I suppose `goreleaser` can be installed outside. As a tool, does it have to be registered as to be like "part" of the project? It's an helper, right?

I mean, I don't list the lib used to build `make` as part of the project.

## building dev env instruction

It was section in the README for that purpose. It was to detail the command to run for building it. (Including an external Golang link you removed I think, to setup the Go workspace).
In fact, at the time I built `doctops` Go was really easy to install with not so bloated stuff to install all around.

And I gave instruction to do it, even for non-Go developer because of the very low dependencies of the project. 

1. Install Go (obviously) and described by Golang page directly, so not repeated in our doc. (tgz extract in /usr/local) + GOPATH + PATH GOBIN
2. install our external tool, was mostly Go `govvv` and such, etc. `make install_builddep`
4. `make`
5. nothing more

the `docopts` binary was here.

There's also a small Dockerfile for experimenting: [./docker/debian-docopts.Dockerfile](https://github.com/docopt/docopts/blob/master/docker/debian-docopts.Dockerfile)

For playing around. (only building `docopts`)

 ## releasing

Releasing was a much more complex step as you lived yourself.

It's a separated process from the build. And it should be reserved (dedicated) to people wanting to publish to github release (unless I finally work on #22, that's another learning curve on my plan).
Basically only me actually.

Because my hacker and automation mindset, I did automate boring repetitive tasks. 
So I finally created `deploy.sh` and moved it to [its own repository](https://github.com/opensource-expert/deploy.sh).

The releasing process is described here: [docs/release.md](https://github.com/docopt/docopts/blob/master/docs/release.md)

 ## what are the required action for building

So as you did an amazing job porting my hack to `goreleaser` :+1:  I want to learn `goreleaser` how it replaces `deploy.sh` but also the version building (at build stage). 

So let's focus first on the build. I will do it you from your branch and produce some updated instructions based on what I will learn.

Then I will try to deploy, but it looks like very similar of what I was doing (+ over than 170 dependencies :yum:  )

Reply:
## `goreleaser` deps

> But are they mandatory to be downloaded just for the purpose of running `goreleaser`?

Yes, as you found out by running `go get github.com/goreleaser/goreleaser@latest` (or `go mod tidy` etc), they're all necessary for `goreleaser` and only `goreleaser`.

### Go modules

> In fact, I started coding Go as a casual coder transposing my experience from other language. But at the time I started Go module was not made that way or not at all.

Yea Go modules didn't exist I think until beta releases of Go `1.11` IIRC. Before that, there was no single dependency management system in Go. Now Go modules standard and built-in, thankfully.

I don't write that much Go myself, so it seemed like you might know the ecosystem better than me at first haha 😅  
In particular, Go workspaces themselves were more complicated than Go modules. So if you were familiar with workspaces, I thought you might have seen modules by now.

> So yes I'm not so familiar with module. I will learn.

No worries though! Go modules are pretty straightforward and simpler than workspaces, so it seems like you've mostly figured it out already!

### tool dependency

> But I suppose, that they're probably not needed to be listed as part of the project? I suppose `goreleaser` can be installed outside. As a tool, does it have to be registered as to be like "part" of the project? It's an helper, right?
> 
> I mean, I don't list the lib used to build `make` as part of the project.

So it is a tool dependency, that's why I put it in [`tools.go`](https://github.com/docopt/docopts/pull/65/files#diff-ce07b69075d5f4237197bb90b721f2b357c8567a1a8945068239b32e57d85552R8).

In other languages, such as JavaScript, [`devDependencies`](https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file) can be listed separately in the package manifest. In JS, `devDependencies` are only installed when you're working on the package itself. If you're just using the package as a dependency, e.g. `import jsreleaser from "jsreleaser";`, the `devDependencies` won't be installed, only `dependencies` will be.

Go modules do not seem to have a "formal" separation of these yet. Per [the Go issue I linked](https://github.com/golang/go/issues/25922#issuecomment-1038394599) above, `tools.go` is a current convention to list tool dependencies that aren't needed when importing, similar to `devDependencies` in JS.

JS's `devDependencies` also make it into JS's `package-lock.json` lockfile, similar to the `go.sum`.
So having `goreleaser` in `go.mod` and `go.sum` follows Go's current conventions. This ensures that anyone developing `docopts` -- in this case you and I -- has the same version of all dependencies, including indirect development dependencies.

#### `go.mod` nuances

As you found out, if you run `go get github.com/goreleaser/goreleaser@latest`, it'll automatically be added to the `go.mod` as an `// indirect` dependency. So it's meant to be listed there. Even if you remove it, it'll get added back as soon as it's installed.

## Building vs. Releasing

> In fact, at the time I built `doctops` Go was really easy to install with not so bloated stuff to install all around.

Agreed I'd like to keep that simple as well. I wasn't sure how to do so when I first made this PR, but I think I may have figured it out now (see next section).

> 1. Install Go (obviously) and described by Golang page directly, so not repeated in our doc. (tgz extract in /usr/local) + GOPATH + PATH GOBIN
> 1. install our external tool, was mostly Go `govvv` and such, etc. `make install_builddep`
> 1. `make`
> 1. nothing more

I should say, the installation actually simpler now, just a bunch of deps get added that aren't used.

> 1. Install Go
> 1. install external tools, `go mod tidy`
> 1. `make`
> 1. nothing more

The steps are simpler, but it is bloated unfortunately

### Releasing

> And it should be reserved (dedicated) to people wanting to publish to github release (unless I finally work on #22, that's another learning curve on my plan).

Agreed.

> The releasing process is described here: [docs/release.md](https://github.com/docopt/docopts/blob/master/docs/release.md)

Ah, I need to modify this doc too with the new process with `goreleaser`. I'll add that change to this PR.

### Building _only_

> ## what are the required action for building

I need to test this a bit more, but apparently Go modules are a lot more intelligent than I first thought.
As I was building the Homebrew package, I realized you can actually just run `go build` _without_ running `go mod tidy` first. And then it _only_ detects that `docopt-go` is a dependency (as it's the only dep in `docopts.go`) and _only_ installs it.

So I think I can remove the `go mod tidy` statement from the `README.md` for individual builds and leave it only for releases.

That will make it easy to build from source without needing `goreleaser` and all of its bloated deps! The only caveat is that the `ldflags` won't be set, but they weren't in the original directions either (just `go build`, with no flags).

Reply:
> I realized you can actually just run `go build` _without_ running `go mod tidy` first. And then it _only_ detects that `docopt-go` is a dependency (as it's the only dep in `docopts.go`) and _only_ installs it.

Yep, this works!

```bash
❯ go clean -modcache
❯ go build
go: downloading github.com/docopt/docopt-go v0.0.0-20180111231733-ee0de3bc6815
```

Only downloads the one necessary dep!

> So I think I can remove the `go mod tidy` statement from the `README.md` for individual builds and leave it only for releases.

Will proceed with this then 🙂

Reply:
Hello, I'm back after a while... (refresh in progress...) ⏳

## Building only

So effectively it works but  won't feed the versioning variables.

```
go build
```

I did some progress, I installed `goreleaser` finally  (from their [doc](https://goreleaser.com/install/#go-install))

```
go install github.com/goreleaser/goreleaser@latest
```

I did update my branch:

```
git branch --set-upstream-to=agilgur5/build-with-goreleaser  agilgur5-build-with-goreleaser
```

(with some conflict may be the one we saw earlier)

So yes I confirm it builds (without variable version updated) which is not what I'm expecting while running `make`.

`$ git diff HEAD master Makefile`

```diff
+# govvv define main.Version with the contents of ./VERSION file, if exists
+BUILD_FLAGS=$(shell ./get_ldflags.sh)
 docopts: docopts.go Makefile
-   go build -o $@
+ go build -o $@ -ldflags "${BUILD_FLAGS} ${LDFLAGS}"
```

`make` should build the local `docopts` binary with all its flags, without complaining about uncommitted changes, and only one arch local target.

I didn't learn enough about goreleaser to ride it my way, it seems it would look like:

```
GOVERSION=$(go version) goreleaser build --single-target --snapshot --rm-dist -o docopts
```
but it still compile all targets and don't left a valid `docopts` in the top folder. 
I will seek for that later.

So it means that Makefile first target `doctops`, should figure out to force gorelaser to replicate the same as `go build -o $@ -ldflags "${BUILD_FLAGS} ${LDFLAGS}"`

Either garbing built ldflags from some magical `goreleaser --extract-ldflags`  
Or multiple action make moving the target file at the correct position.

No more time for today, talk later.

Reply:
> So effectively it works but won't feed the versioning variables.

Yea, I realize that.

But I think that's a fine trade-off for quick builds and builds from source.
In fact, you _already_ have that in [the README](https://github.com/docopt/docopts/blame/190c0f2a3e8071e2a1b06f0b90f2556c7d079f28/README.md#L385): `go build docopts.go` doesn't provide any versioning information.
Since it was already in the README, that's why I thought it was ok to make [that change](https://github.com/docopt/docopts/pull/65/commits/548ba408daa7d0a89b74e0c0d71c13939bfc18fc).

As this does not change the approach in the README, I think this change is fine.

> ```
> GOVERSION=$(go version) goreleaser build --single-target --snapshot --rm-dist -o docopts
> ```

And yes, if you want all the versioning you can run this. That's what I [previously had](https://github.com/docopt/docopts/pull/65/commits/ae7431ba25986a6c65cd337fcb77320a5c86c45c).

> but it still compile all targets

`make snapshot` that I have in this PR does so. But `make snapshot` doesn't use [`--single-target`](https://goreleaser.com/cmd/goreleaser_build/). We can add another command that does only `--single-target` etc.

> So it means that Makefile first target `doctops`

I can change the default `make` if you want, but you might as well use `make all` at that point -- which is what the README already had as well.

Reply:
## what are the required action for building

README seems a bit outdated at some point. Thanks for bringing me the eyes on that. As I said, it was simple long time ago... :wink:

So lets focus, as being the main builder, and a `vim` user, and a `make` command user. Just `make` should be smart enough to accomplish complex work including knowing what to do.

It comes back from good old C compiling only requires changed dependencies, etc. As you may probably know. So make is very smart. And despite `go build` figures out itself a lot of thing if may be difficult to tell him to take care of `VERSION` file correctly.  

Was done by `govvv` initially 

```Makefile
BUILD_FLAGS=$(shell ./get_ldflags.sh)
docopts: docopts.go Makefile
	go build -o $@ -ldflags "${BUILD_FLAGS} ${LDFLAGS}"
```

So I need a way to get it back. `go build` without argument could be OK for casual building. I checked, there's was no functional testing for `docopts --version`. See below some internal history famous details about why `--version`. It could be tested will full variable expansion during development stage too or when I'm crafting the `--version` message too. So the default Makefile rule should be much more tricky than just a trivial `go build`.

## `docopts --version` gory details

You may have noticed [there's a hack on `--version=<msg>` switch vs `--version`](https://github.com/docopt/docopts/blob/190c0f2a3e8071e2a1b06f0b90f2556c7d079f28/docopts.go#L340) without argument also triggering a "normal" software version output and exit(0). This was kept from backward compatibility of the original author of `docopts`.

I rarely myself use versioning in my own cli but that's part of the cli API of `docopts` itself for now.

So some future unit test may be introduced that the `--version` with or without argument may pass or fail. 

I'm also disagree with the `--help=<msg>` too, may I decide to break it as some version to keep the code more easy to understand without going to exception handler, or such...

## `--single-target` still build all target

For some unknown reason `goreleaser` still build all targets:

```
sylvain@lap43: ~/code/go/src/github.com/docopt/docopts$ make
GOVERSION=$(go version) goreleaser build --single-target --rm-dist --snapshot -o docopts
  • starting build...
  • loading config file                              file=.goreleaser.yaml
  • building only for linux/amd64                    reason=single target is enabled
  • loading environment variables
  • getting and validating git state
    • building...                                    commit=34b0c0ba4c696662087eeb1ef489f96558899951 latest tag=v0.6.4-with-no-mangle-double-dash
    • pipe skipped                                   reason=disabled during snapshot mode
  • parsing tag
  • setting defaults
  • running before hooks
    • running                                        hook=go mod tidy
  • snapshotting
    • building snapshot...                           version=0.6.4-next
  • checking distribution directory
  • loading go mod information
  • build prerequisites
  • writing effective config file
    • writing                                        config=dist/config.yaml
  • building binaries
    • building                                       binary=dist/docopts_linux_amd64_v1/docopts
    • building                                       binary=dist/docopts_darwin_arm64/docopts
    • building                                       binary=dist/docopts_freebsd_amd64_v1/docopts
    • building                                       binary=dist/docopts_linux_386/docopts
    • building                                       binary=dist/docopts_linux_arm_6/docopts
    • building                                       binary=dist/docopts_windows_amd64_v1/docopts.exe
    • building                                       binary=dist/docopts_darwin_amd64_v1/docopts
    • took: 1s
  • storing release metadata
    • writing                                        file=dist/artifacts.json
    • writing                                        file=dist/metadata.json
  • copying binary to "docopts"
  • build succeeded after 1s
```

```
sylvain@lap43: ~/code/go/src/github.com/docopt/docopts$ tree dist
dist
├── artifacts.json
├── config.yaml
├── docopts_darwin_amd64_v1
│   └── docopts
├── docopts_darwin_arm64
│   └── docopts
├── docopts_freebsd_amd64_v1
│   └── docopts
├── docopts_linux_386
│   └── docopts
├── docopts_linux_amd64_v1
│   └── docopts
├── docopts_linux_arm_6
│   └── docopts
├── docopts_windows_amd64_v1
│   └── docopts.exe
└── metadata.json

7 directories, 10 files
```

at least `-o` is honored and a `docopts` amd64 binary is present

```
./docopts --version
docopts 0.6.4-next commit 34b0c0b built at 2022-09-11T08:02:49Z
built from: go version go1.19.1 linux/amd64
Copyleft (Ɔ)  2022 Sylvain Viart (golang version).
Copyright (C) 2013 Vladimir Keleshev, Lari Rasku.
License MIT <http://opensource.org/licenses/MIT>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

may with multiple build target in the `.goreleaser.yaml` and a `--id` on the `--single-target`?

```yaml
# .goreleaser.yaml
builds:
  - main: ./dont/know/the/path/here
    id: "dev"
    binary: docopts
```

I will probably ask on https://github.com/goreleaser/goreleaser/discussions
But I've to read much more about `goreleaser`. 

I've the feeling that `goreleaser`, may be needed to be kept in a sub-folder to not interfere with basic build new requesting release. So moving `tools.go` to a sub-folder etc.

## Makefile change

Do you want me to push that change to your branch or do you prefer me not to mess your code base? Because, as a maintainer, I've right to push on PR branches to collaborate.

```diff
diff --git a/Makefile b/Makefile
index 8cacfec..f45ad29 100644
--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,7 @@ RELEASE_NOTES := $$(awk -v RS='\#\# *|\#\# ' 'NR==2 { print }' CHANGELOG.md)
 # keep docopts: as first target for development
 
 docopts: docopts.go Makefile
-   go build -o $@
+ GOVERSION=$(GOVERSION) goreleaser build --single-target --rm-dist --snapshot -o docopts
 
 install_builddep:
        go mod tidy
@@ -34,7 +34,9 @@ README.md: examples/legacy_bash/rock_hello_world.sh examples/legacy_bash/rock_he
        mv README.tmp README.md
 
 clean:
-   rm -f docopts-* docopts README.tmp dist/*
+ # remove goreleaser dist/ difrectrory recursively
+ rm -rf dist
+ rm -f docopts-* docopts README.tmp
 
 test_release_notes:
        echo "\n## $(RELEASE_NOTES)"
```

Reply:
> `go build` without argument could be OK for casual building. I checked, there's was no functional testing for `docopts --version`

> So some future unit test may be introduced that the `--version` with or without argument may pass or fail.

I think we can leave it as such for casual building. If needed, the `VERSION` file is still there, so `cat VERSION` can probably be added to the `Makefile` too. Could probably do the same for the other `ldflags` too.

> You may have noticed [there's a hack on `--version=<msg>` switch vs `--version`](https://github.com/docopt/docopts/blob/190c0f2a3e8071e2a1b06f0b90f2556c7d079f28/docopts.go#L340) without argument also triggering a "normal" software version output and exit(0). This was kept from backward compatibility of the original author of `docopts`.

> I'm also disagree with the `--help=<msg>` too, may I decide to break it as some version to keep the code more easy to understand without going to exception handler, or such...

Yea I did notice that and read about it.  Agreed that it's not ideal but seems like an ok workaround for now

> For some unknown reason `goreleaser` still build all targets:

Think we've figured this out in https://github.com/goreleaser/goreleaser/discussions/3371#discussioncomment-3656182 / https://github.com/goreleaser/goreleaser/pull/3381 -- we found a bug!

> Do you want me to push that change to your branch or do you prefer me not to mess your code base? Because, as a maintainer, I've right to push on PR branches to collaborate.

Feel free to push to my branch! For reference, the PR author actually has a checkbox to allow or deny those permissions to maintainers. By default it is allowed, and as a maintainer myself, it seems to be generally accepted to do so -- no need to ask!

I noticed `make clean` needed changing too recently, and with the fixes to `--single-target` and your requests for the default `make`, those changes would be needed too. So all the changes make sense to me 👍